client/components/svg-canvas.jsx

```diff
@@ -60,9 +59,25 @@ export default class SVGCanvas extends React.Component {
     this.handleTouchEnd = this.handleTouchEnd.bind(this);
     this.handleKeydown = this.handleKeydown.bind(this);
     this.handleTextboxModalSubmit = this.handleTextboxModalSubmit.bind(this);
+    this.saveElementsToDB = this.saveElementsToDB.bind(this);
   }
 
-  componentDidUpdate(prevProps) {
+  componentDidMount() {
+    // Simple saving mechanism, saves elements array to DB every 5 seconds as long as
+    // it's not empty
+    // the user is not currently adding a new element
+    // the length of the elements array is different than before
+    let previousElementsLength = this.state.elements.length;
+    setInterval(() => {
+      if (this.state.elements.length === 0 || this.state.currentElementId !== null || this.state.elements.length === previousElementsLength) {
+        return;
+      }
+      previousElementsLength = this.state.elements.length;
+      this.saveElementsToDB();
+    }, 5000);
+  }
```

Comment:
Typically we try not to set up intervals for stuff like this as it could result in needless network transfers. Instead, leverage `componentDidUpdate` and a technique called "debouncing". Debouncing is when it's safe to spam a function call because it ignores all calls but its last, given some quiet timeout threshold. If I'm not mistaken, it's an electrical engineering term that refers to using capacitors to absorb spikes/stutters in voltage when switching circuits on or off. Closures to the rescue.

```js
function debounce(fn, delay) {
  let timeout = null
  return function (...args) {
    clearTimeout(timeout)
    timeout = setTimeout(fn, delay, ...args)
  }
}
```

you can store a bound, debounced version of your saveElements method on your instance in the constructor.

```js
this.debouncedSaveElementsToDB = debounce(this.saveElementsToDB.bind(this))
```

In `componentDidUpdate`, auto-save the drawing on each update using your debounced method, assuming the drawing isn't empty and has actually changed since the last update.

Reply:
Wow, that one is going into the flashcards 🤯